src/libsyntax/tokenstream.rs

```diff
@@ -820,98 +929,108 @@ impl Index<usize> for TokenStream {
     type Output = TokenTree;
 
     fn index(&self, index: usize) -> &TokenTree {
-        Index::index(&**self, index)
+      if self.len() <= index { panic!("Index {} too large for {:?}", index, self); }
+      match *self {
+        TokenStream::Empty                            => panic!("Invalid index"),
+        TokenStream::Leaf { ref tts, .. }             => tts.get(index).unwrap(),
+        TokenStream::SubLeaf { ref tts, offset, .. }  => tts.get(index + offset).unwrap(),
+        TokenStream::Node { ref left, ref right, .. } => {
+          let left_len = left.len();
+          if index < left_len {
+            Index::index(&**left, index)
+          } else {
+            Index::index(&**right, index - left_len)
+          }
+        }
+      }
     }
 }
 
-impl ops::Index<ops::Range<usize>> for TokenStream {
-    type Output = TokenSlice;
 
-    fn index(&self, index: ops::Range<usize>) -> &TokenSlice {
-        Index::index(&**self, index)
-    }
-}
-
-impl ops::Index<ops::RangeTo<usize>> for TokenStream {
-    type Output = TokenSlice;
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ast;
+    use syntax_pos::{Span, BytePos, NO_EXPANSION, DUMMY_SP};
+    use parse::token::{self, str_to_ident, Token, Lit};
+    use util::parser_testing::string_to_tts;
+    use std::rc::Rc;
 
-    fn index(&self, index: ops::RangeTo<usize>) -> &TokenSlice {
-        Index::index(&**self, index)
+    fn sp(a: u32, b: u32) -> Span {
+        Span {
+            lo: BytePos(a),
+            hi: BytePos(b),
+            expn_id: NO_EXPANSION,
+        }
     }
-}
 
-impl ops::Index<ops::RangeFrom<usize>> for TokenStream {
-    type Output = TokenSlice;
-
-    fn index(&self, index: ops::RangeFrom<usize>) -> &TokenSlice {
-        Index::index(&**self, index)
+    #[test]
+    fn test_concat() {
+      let test_res = TokenStream::from_tts(string_to_tts("foo::bar::baz".to_string()));
```

Comment:
Could use a `&str -> TokenSream` conversion method and/or a `FromStr` implementation.